cache/cache.go

```diff
@@ -13,3 +13,235 @@
 // limitations under the License.
 
 package cache
+
+import (
+	"context"
+	"sort"
+	"strings"
+	"sync/atomic"
+
+	rpctypes "go.etcd.io/etcd/api/v3/v3rpc/rpctypes"
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+var ErrCompacted = rpctypes.ErrCompacted
+
+type Stream struct{ Ch <-chan *clientv3.Event }
+
+// Cache buffers a single etcd Watch for a given key‐prefix and fan‑outs local watchers.
+// It is created internally by ShardedCache; callers typically use ShardedCache instead.
+type Cache struct {
+	prefix             string // keys this shard is responsible for, may be "" (root)
+	cfg                Config
+	etcd               *clientv3.Client
+	hist               History
+	mux                *demux
+	ctx                context.Context
+	cancel             context.CancelFunc
+	latestCompactedRev int64         // updated atomically
+	ready              chan struct{} // closed when initialLoad finishes
+}
+
+// NewWithPrefix builds a cache shard that watches only the requested prefix.
+// For the root cache pass "".
+func NewWithPrefix(cli *clientv3.Client, prefix string, opts ...Option) (*Cache, error) {
+	cfg := defaultConfig()
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// choose history implementation
+	ringBuffer := newRingBuffer(cfg.BufferEntries)
+
+	cache := &Cache{
+		prefix: prefix,
+		cfg:    cfg,
+		etcd:   cli,
+		hist:   ringBuffer,
+		mux:    newDemux(),
+		ctx:    ctx,
+		cancel: cancel,
+		ready:  make(chan struct{}),
+	}
+
+	// initial chatch up
+	if err := cache.initialLoad(); err != nil {
+		cancel()
+		return nil, err
+	}
+	close(cache.ready)
+
+	// upstream watch goroutine
+	serveWatchEvents(
+		ctx,
+		cli,
+		prefix,
+		cache.mux.broadcast,
+		cache.mux,
+		cache.hist,
+		func(rev int64) { atomic.StoreInt64(&cache.latestCompactedRev, rev) },
+		cfg.InitialBackoff,
+		cfg.MaxBackoff,
+	)
+
+	return cache, nil
+}
+
+// initialLoad ranges over the prefix and seeds the ring.
+func (c *Cache) initialLoad() error {
+	opts := []clientv3.OpOption{clientv3.WithPrefix()}
+	resp, err := c.etcd.Get(c.ctx, c.prefix, opts...)
+	if err != nil {
+		return err
+	}
+	for _, kv := range resp.Kvs {
+		event := &clientv3.Event{
+			Type: clientv3.EventTypePut,
+			Kv:   kv,
+		}
+		c.hist.Append(event)
+	}
+	return nil
+}
+
+func (c *Cache) Watch(ctx context.Context, key string, opts ...clientv3.OpOption) clientv3.WatchChan {
+	select {
+	case <-c.ready:
+	case <-ctx.Done():
+		emptyWatchChan := make(chan clientv3.WatchResponse)
+		close(emptyWatchChan)
```

Comment:
Think you need to push a event to represent that watch was cancelled.

Something like `&WatchResponse{Canceled: true, closeErr: err}`, however looks like `closeErr` is a private field, let's skip it for now.